### PR TITLE
Stacksafety refactors

### PR DESCRIPTION
**Pull request by ArturGajowy**
## Overview
Title says it all :)

One notable change is 
1974667 Use `.serializedSize` instead of serializing to get message storage cost

This has a potential of improving performance of rspace (or, rather, our usages of it).

### JIRA issue:
RHOL-664

### Checklist
- [x] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### CODE REVIEW

File: models/src/main/scala/coop/rchain/models/ParMap.scala [hunk 1]
```diff
@@ -15,11 +15,12 @@ case class ParMap(
 ) {
 
   override def equals(o: scala.Any): Boolean = o match {
-    case parMap: ParMap => this.ps == parMap.ps && this.remainder == parMap.remainder
-    case _              => false
+    case parMap: ParMap =>
+      this.ps == parMap.ps && this.remainder == parMap.remainder && this.connectiveUsed == parMap.connectiveUsed
```
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply models/src/main/scala/coop/rchain/models/ParMap.scala]: > If connectiveUsed were different it means we have a bug.

Agreed. Then it makes even more sense to make the equals consistent with other variants of the AST, because comparing it can lead to exposing a bug. Ignoring it can't.

File: models/src/test/scala/coop/rchain/models/testImplicits.scala [hunk 2]
```diff
@@ -27,7 +28,42 @@ object testImplicits {
   implicit def coeval[A: Arbitrary]: Arbitrary[Coeval[A]] =
     Arbitrary(Arbitrary.arbitrary[A].map(a => Coeval.delay(a)))
 
-  implicit val arbPar = implicitly[Arbitrary[Par]]
```
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply models/src/test/scala/coop/rchain/models/testImplicits.scala]: It was, the others are added for a bigger confidence level, so that we can test on a tree starting from whatever AST variant we want

File: models/src/test/scala/coop/rchain/models/RhoTypesTest.scala [hunk 3]
```diff
@@ -5,24 +5,39 @@ import coop.rchain.models.BitSetBytesMapper._
...
 import scala.collection.immutable.BitSet
+import scala.reflect.ClassTag
 
 class RhoTypesTest extends FlatSpec with PropertyChecks with Matchers {
 
-  def rtt[T](msg: T)(implicit serializeInstance: Serialize[T]): Either[Throwable, T] = {
-    val bytes = serializeInstance.encode(msg)
-    serializeInstance.decode(bytes)
-  }
+  //FIXME crank that up and fix the resulting errors
```
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply models/src/test/scala/coop/rchain/models/RhoTypesTest.scala]: One of them is fixed by this PR, I anticipate others. Hope I'm wrong.

[the remark this answers is not in the document]
Comment (goral09) [reply models/src/test/scala/coop/rchain/models/RhoTypesTest.scala]: What was the error?

[the remark this answers is not in the document]
Comment (ArturGajowy) [reply models/src/test/scala/coop/rchain/models/RhoTypesTest.scala]: The one fixed by:
https://github.com/rchain/rchain/pull/1606/commits/936bd288b8ed54a00bd8e2e146eed0c50db09875

TL;DR there was a possibility to create an expression that would make the normalizer barf with MatchError. Using generative tests requires handling all representable states, even if they're invalid (in which case they shouldn't be representable in the first place).

Notice this only surfaced after I introduced an explicit test for Expr and cranked up the maxSize and minSuccesses to current value.

[the remark this answers is not in the document]
Comment (goral09) [reply models/src/test/scala/coop/rchain/models/RhoTypesTest.scala]: Hmm OK. We wouldn't have this problem if we handled all the cases in the `match` (make warnings as errors).
